File: android/src/main/java/expo/modules/viewportmetrics/ViewportMetricsModule.kt
package expo.modules.viewportmetrics

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class ViewportMetricsModule : Module() {
  private val aggregator: ViewportMetricsAggregator
    get() = ViewportMetricsAggregatorRegistry.get(appContext)

  override fun definition() = ModuleDefinition {
    Name("ViewportMetrics")

    Events("onSnapshot")

    OnCreate {
      aggregator.setModuleEmitter { snapshot ->
        sendEvent("onSnapshot", snapshot)
      }
      aggregator.start()
    }

    OnDestroy {
      aggregator.setModuleEmitter(null)
      aggregator.stop()
    }

    OnStartObserving("onSnapshot") {
      aggregator.setJsObserving(true)
    }

    OnStopObserving("onSnapshot") {
      aggregator.setJsObserving(false)
    }

    OnActivityEntersForeground {
      aggregator.start()
      aggregator.markDirty("activityForeground")
    }

    OnActivityEntersBackground {
      aggregator.pauseSensors()
    }

    Function("getSnapshot") {
      aggregator.currentSnapshot()
    }

    View(ViewportMetricsView::class) {
      Events("onSnapshot")
    }
  }
}

File: ios/ViewportMetricsModule.swift
import ExpoModulesCore

public class ViewportMetricsModule: Module {
  private var aggregator: ViewportMetricsAggregator? {
    guard let appContext else {
      return nil
    }
    return ViewportMetricsAggregatorRegistry.shared.get(appContext: appContext)
  }

  public func definition() -> ModuleDefinition {
    Name("ViewportMetrics")

    Events("onSnapshot")

    OnCreate {
      self.aggregator?.setModuleEmitter { [weak self] snapshot in
        self?.sendEvent("onSnapshot", snapshot)
      }
      self.aggregator?.start()
    }

    OnDestroy {
      self.aggregator?.setModuleEmitter(nil)
      self.aggregator?.stop()
    }

    OnStartObserving("onSnapshot") {
      self.aggregator?.setJsObserving(true)
    }

    OnStopObserving("onSnapshot") {
      self.aggregator?.setJsObserving(false)
    }

    Function("getSnapshot") {
      return self.aggregator?.currentSnapshot() ?? ViewportMetricsAggregator.fallbackSnapshot()
    }

    View(ViewportMetricsView.self) {
      Events("onSnapshot")
    }
  }
}
